Add option to fail on errors

manuelgrabowski opened the pull request:
I need our pipeline to fail in case of errors from CodeSniffer, so no one could accidentally merge code that does not fit PSR-2. Currently, the plugin only reports the errors, but Danger does not fail if there are any.

I've added a new option called `fail_on_error`. After the markdown report is sent, I check if the option is enabled and if there have been any errors and tell Danger to fail the build in that case. (Will only fail in case the `--fail-on-errors` setting for Danger is used, of course.) I documented the new option in the usage section of the Readme and also reformatted it slightly to make it more readable.

I'm afraid I am not 100% certain if this actually works, because I haven't found any proper documention for writing Danger plugins. I've looked into another plugin to see how they fail builds in case of errors. For example here: https://github.com/dblock/danger-changelog/blob/5fa7295a3fc4a1b749818333c9e3c5bbd7683812/lib/changelog/plugin.rb#L91

I also don't have much experience with Ruby and Rubygems, so I just copied what you did on the last version update in order to hopefully save you some work. :) But please let me know how I could have done anything better, I'm happy to learn more in this area!

ymhuang0808 commented:
Hi @manuelgrabowski, thank you for the PR. It makes better for the Danger plugin and the code looks good. I would review it later.

manuelgrabowski commented:
Hi @ymhuang0808, did you get a chance to look at it yet? :)

If it's okay with you, I'd do some more work on the plugin to solve some other little issues we noticed when using it on our projects. I'd like to get this one live before continuing, though.

ymhuang0808 commented on lib/php_codesniffer/plugin.rb:
```diff
@@ -30,6 +30,11 @@ class DangerPhpCodesniffer < Plugin
+    # An attribute for failing on errors
+    # In case phpcs finds errors, Danger will fail the pipeline
+    # @return [Boolean]
+    attr_accessor :failonerror
```
According to the naming convention in Ruby, it should be written in lowercase and used by underscores to separate words.

manuelgrabowski replied:
Changed it to `fail_on_error`, thanks for letting me know! :)

ymhuang0808 commented on lib/php_codesniffer/plugin.rb:
```diff
@@ -63,6 +68,9 @@ def exec
       markdown generate_summary_markdown summary
       markdown report
+      if failonerror && summary["errors"] > 0
+        messaging.failure("There are #{summary["errors"]} errors that need to be resolved.")
```
It may use  a simple function `fail()`. 
https://github.com/danger/danger/blob/master/Dangerfile#L40

manuelgrabowski replied:
Switched to `fail()` now 👍

manuelgrabowski commented:
Hi @ymhuang0808 – no worries, thanks for getting back to me now and thanks for the feedback! :)

I've changed the things you commented on. I also added information about `--fail-on-errors`. You're completely right, of course. We've always been running Danger with that setting, so I didn't think about it when doing the changes.

ymhuang0808 commented:
@manuelgrabowski I appreciated for your contribution. In the PR, you made the documentation more great and clear. I merged the PR and published on Rubygems.

manuelgrabowski commented:
Thanks @ymhuang0808 :) I've already deployed the new version into our pipeline and it works really well, our developers will be very happy! As I said I will probably open a few more PRs in the future, thanks again for providing helpful feedback and helping me getting the changes in, really appreciate it!
